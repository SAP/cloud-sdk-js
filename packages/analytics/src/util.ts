/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import crypto from 'crypto';

/**
 * Generate a random salt.
 * @returns A random salt.
 * @hidden
 */
export function randomSalt(): string {
  // 1 byte will produce two bytes in hex string
  // 32 random bytes generate a hex string of the same bitlength as SHA-256
  return crypto.randomBytes(32).toString('hex');
}

/**
 * Create a hash for the given value.
 * @param value - The value to hash.
 * @returns The hashed value.
 * @hidden
 */
export function hash(value: string): string {
  return crypto
    .createHash('sha256')
    .update(value)
    .digest('hex');
}
