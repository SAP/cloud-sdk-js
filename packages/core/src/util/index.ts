export * from './jwt';
export * from './name-converter';
export * from './properties-util';
export * from './proxy-util';
export * from './remove-slashes';
