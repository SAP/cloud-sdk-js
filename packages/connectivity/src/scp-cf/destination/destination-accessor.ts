import { createLogger, ErrorWithCause } from '@sap-cloud-sdk/util';
import { exchangeToken, isTokenExchangeEnabled } from '../identity-service';
import { parseSubdomain } from '../subdomain-replacer';
import {
  DestinationOrFetchOptions,
  sanitizeDestination,
  toDestinationNameUrl
} from './destination';
import { Destination } from './destination-service-types';
import { searchEnvVariablesForDestination } from './destination-from-env';
import {
  DestinationForServiceBindingOptions,
  searchServiceBindingForDestination
} from './destination-from-vcap';
import {
  getDestinationFromDestinationService,
  DestinationFromServiceRetriever,
  getDestinationServiceCredentials
} from './destination-from-service';
import {
  DestinationFetchOptions,
  isDestinationFetchOptions,
  AllDestinationOptions,
  DestinationWithoutToken
} from './destination-accessor-types';
import { searchRegisteredDestination } from './destination-from-registration';
import {
  fetchInstanceDestinations,
  fetchSubaccountDestinations
} from './destination-service';

const logger = createLogger({
  package: 'connectivity',
  messageContext: 'destination-accessor'
});

/**
 * Returns the parameter if it is a destination, calls {@link getDestination} otherwise (which will try to fetch the destination
 * from the Cloud Foundry destination service).
 *
 * Fetching a destination requires:
 * - A binding to exactly one XSUAA service instance with service plan "application".
 * - A binding to a destination service instance.
 *
 * If either of the prerequisites is not met or one of the services returns an error, this function will either throw an error or return a promise that rejects.
 * @param destination - A destination or the necessary parameters to fetch one.
 * @returns A promise resolving to the requested destination on success.
 */
export async function useOrFetchDestination(
  destination: DestinationOrFetchOptions
): Promise<Destination | null> {
  return isDestinationFetchOptions(destination)
    ? getDestination(destination)
    : sanitizeDestination(destination);
}

/**
 * Resolve a destination by the following steps:
 * 1. Call [[useOrFetchDestination]]
 * 2. Throw an error, when the resulting destination from the previous step is falsy
 * 3. Return the checked destination.
 * @param destination - A destination or the necessary parameters to fetch one.
 * @returns A promise resolving to the requested destination on success.
 * @internal
 */
export async function resolveDestination(
  destination: DestinationOrFetchOptions
): Promise<Destination> {
  const resolvedDestination = await useOrFetchDestination(destination).catch(
    error => {
      throw new ErrorWithCause('Failed to load destination.', error);
    }
  );
  if (!resolvedDestination) {
    throw Error(
      `Failed to resolve the destination '${toDestinationNameUrl(
        destination
      )}'.`
    );
  }
  return resolvedDestination;
}

/**
 * Builds a destination from one of three sources (in the given order):
 * - from the environment variable "destinations".
 * - from service bindings.
 * - from the destination service.
 *
 * If you want to get a destination only from a specific source, use the corresponding function directly
 * (`getDestinationFromEnvByName`, `destinationForServiceBinding`, `getDestinationFromDestinationService`).
 * @param options - The options to retrieve the destination.
 * @returns A promise returning the requested destination on success.
 */
export async function getDestination(
  options: DestinationFetchOptions & DestinationForServiceBindingOptions
): Promise<Destination | null> {
  const destination =
    searchEnvVariablesForDestination(options) ||
    (await searchRegisteredDestination(options)) ||
    (await searchServiceBindingForDestination(options)) ||
    (await getDestinationFromDestinationService(options));
  return destination;
}

/**
 * Creates comprehensive log messages from a destinations array and their origin.
 * @param origin - Origin of the destination.
 * @param destinations - Array of destinations.
 * @returns Logs of the retrival of destinations.
 */
function createDestinationFetchLogs(
  origin: string,
  destinations: Destination[]
): string {
  return destinations.reduce(
    (prevLogMessages, currentDestination) =>
      prevLogMessages +
      `Retrieving ${origin} destination: ${currentDestination.name}.\n`,
    ''
  );
}

/**
 * Fetches all destinations from the destination service which match the token.
 * With a subscriber token it fetches all subscriber destinations, otherwise all provider destinations.
 * @param options - The {@link AllDestinationOptions | options} to fetch all destinations.
 * @returns A promise of an array of all destinations without authTokens from the destination service, on success.
 */
export async function getAllDestinationsFromDestinationService(
  options: AllDestinationOptions = {}
): Promise<DestinationWithoutToken[]> {
  logger.debug(
    'Attempting to retrieve all destinations from destination service.'
  );
  if (isTokenExchangeEnabled(options)) {
    options.jwt = await exchangeToken(options);
  }

  const token =
    (await DestinationFromServiceRetriever.getSubscriberToken(options))
      ?.serviceJwt ||
    (await DestinationFromServiceRetriever.getProviderServiceToken(options));

  const destinationServiceUri = getDestinationServiceCredentials().uri;
  const accountName = parseSubdomain(token.decoded.iss!);
  logger.debug(
    `Retrieving all destinations for account: "${accountName}" from destination service.`
  );

  const [instance, subaccount] = await Promise.all([
    fetchInstanceDestinations(destinationServiceUri, token.encoded, options),
    fetchSubaccountDestinations(destinationServiceUri, token.encoded, options)
  ]);

  const loggerMessage =
    createDestinationFetchLogs('instance', instance) +
    createDestinationFetchLogs('subaccount', subaccount);

  logger.debug(loggerMessage);

  const allDestinations = instance.concat(subaccount);

  if (allDestinations && allDestinations.length !== 0) {
    logger.debug(
      `Successfully retrieved all destinations for account: "${accountName}" from destination service.`
    );
  } else {
    logger.debug("Didn't receive any destinations from destination service.");
    return [];
  }

  const allDestinationsWithoutToken: DestinationWithoutToken[] =
    allDestinations.map(destination => {
      delete destination.authTokens;
      return destination;
    });

  return allDestinationsWithoutToken;
}
