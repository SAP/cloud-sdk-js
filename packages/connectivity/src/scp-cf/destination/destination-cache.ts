import { createLogger, first } from '@sap-cloud-sdk/util';
import { Cache, CacheEntry } from '../cache';
import { tenantId } from '../tenant';
import { userId } from '../user';
import { JwtPayload } from '../jsonwebtoken-type';
import { Destination } from './destination-service-types';
import { DestinationsByType } from './destination-accessor-types';

const logger = createLogger({
  package: 'connectivity',
  messageContext: 'destination-cache'
});

/**
 * Enumerator that selects the isolation type of destination in cache.
 * The used isolation strategy is either `Tenant` or `Tenant_User` because we want to get results for subaccount and provider tenants which rules out no-isolation or user isolation.
 */
export enum IsolationStrategy {
  Tenant = 'Tenant',
  Tenant_User = 'TenantUser'
}

/**
 * Interface to implement custom destination caching.
 * To set a custom implementation, call method {@link setDestinationCache} and pass the cache instance.
 */
export interface DestinationCacheInterface {
  hasKey(key: string): Promise<boolean>;
  get(key: string | undefined): Promise<Destination | undefined>;
  set(key: string | undefined, item: CacheEntry<Destination>): Promise<void>;
  clear(): Promise<void>;
}

/**
 * @internal
 * This wrapper class wraps methods of {@link Cache} class as asynchronous methods.
 */
export class DefaultDestinationCache implements DestinationCacheInterface {
  cache: Cache<Destination>;

  constructor(validityTimeInMs?: number) {
    this.cache = new Cache<Destination>(validityTimeInMs);
  }

  /**
   * Specifies whether an entry with a given key is defined in cache.
   * @param key - The entry's key.
   * @returns A boolean value that indicates whether the entry exists in cache.
   */
  async hasKey(key: string): Promise<boolean> {
    return this.cache.hasKey(key);
  }

  /**
   * Getter of cached entries.
   * @param key - The key of the entry to retrieve.
   * @returns The corresponding entry to the provided key if it is still valid, returns `undefined` otherwise.
   */
  async get(key: string | undefined): Promise<Destination | undefined> {
    return this.cache.get(key);
  }

  /**
   * Setter of entries in cache.
   * @param key - The entry's key.
   * @param item - The entry to cache.
   */
  async set(
    key: string | undefined,
    item: CacheEntry<Destination>
  ): Promise<void> {
    return this.cache.set(key, item);
  }

  /**
   * Clear all cached items.
   */
  async clear(): Promise<void> {
    return this.cache.clear();
  }
}

/**
 * @internal
 */
export interface DestinationCacheType {
  /**
   * @internal
   */
  retrieveDestinationFromCache: (
    decodedJwt: Record<string, any>,
    name: string,
    isolation: IsolationStrategy
  ) => Promise<Destination | undefined>;
  /**
   * @internal
   */
  cacheRetrievedDestination: (
    decodedJwt: Record<string, any>,
    destination: Destination,
    isolation: IsolationStrategy
  ) => Promise<void>;
  /**
   * @internal
   */
  cacheRetrievedDestinations: (
    decodedJwt: Record<string, any>,
    retrievedDestinations: DestinationsByType,
    isolation: IsolationStrategy
  ) => Promise<void>;
  /**
   * @internal
   */
  clear: () => Promise<void>;
  /**
   * @internal
   */
  getCacheInstance: () => DestinationCacheInterface;
}

/**
 * DestinationCache constructor.
 * @param cache - Cache object which is used in DestiantionCache
 * @returns A destination cache object.
 * @internal
 */
export const DestinationCache = (
  cache: DestinationCacheInterface = new DefaultDestinationCache(300000)
): DestinationCacheType => ({
  retrieveDestinationFromCache: async (
    decodedJwt: Record<string, any>,
    name: string,
    isolation: IsolationStrategy
  ): Promise<Destination | undefined> =>
    cache.get(getDestinationCacheKey(decodedJwt, name, isolation)),
  cacheRetrievedDestination: async (
    decodedJwt: Record<string, any>,
    destination: Destination,
    isolation: IsolationStrategy
  ): Promise<void> => {
    cacheRetrievedDestination(decodedJwt, destination, isolation, cache);
  },
  cacheRetrievedDestinations: async (
    decodedJwt: Record<string, any>,
    retrievedDestinations: DestinationsByType,
    isolation: IsolationStrategy
  ): Promise<void> => {
    retrievedDestinations.subaccount.forEach(dest =>
      cacheRetrievedDestination(decodedJwt, dest, isolation, cache)
    );
    retrievedDestinations.instance.forEach(dest =>
      cacheRetrievedDestination(decodedJwt, dest, isolation, cache)
    );
  },
  clear: async (): Promise<void> => {
    cache.clear();
  },
  getCacheInstance: () => cache
});

/**
 * Calculates a cache key based on the jwt and destination name for the given isolation strategy.
 * Cache keys for strategies are non-overlapping, i.e. using a cache key for strategy {@link IsolationStrategy.Tenant}
 * will not result in a cache hit for a destination that has been cached with strategy {@link IsolationStrategy.Tenant_User}.
 * @param decodedJwt - The decoded JWT of the current request.
 * @param destinationName - The name of the destination.
 * @param isolationStrategy - The strategy used to isolate cache entries.
 * @returns The cache key.
 * @internal
 */
export function getDestinationCacheKey(
  decodedJwt: Record<string, any>,
  destinationName: string,
  isolationStrategy = IsolationStrategy.Tenant_User
): string | undefined {
  const tenant = tenantId(decodedJwt);
  const user = userId(decodedJwt);
  switch (isolationStrategy) {
    case IsolationStrategy.Tenant:
      if (tenant) {
        return `${tenant}::${destinationName}`;
      }
      logger.warn(
        `Cannot get cache key. Isolation strategy ${isolationStrategy} is used, but tenant id is undefined.`
      );
      return;
    case IsolationStrategy.Tenant_User:
      if (tenant && user) {
        return `${user}:${tenant}:${destinationName}`;
      }
      logger.warn(
        `Cannot get cache key. Isolation strategy ${isolationStrategy} is used, but tenant id or user id is undefined.`
      );
      return;
    default:
      logger.warn(
        `Cannot get cache key. Isolation strategy ${isolationStrategy} is not supported.`
      );
      return;
  }
}

async function cacheRetrievedDestination<T extends DestinationCacheInterface>(
  decodedJwt: Record<string, any>,
  destination: Destination,
  isolation: IsolationStrategy,
  cache: T
): Promise<void> {
  if (!destination.name) {
    throw new Error('The destination name is undefined.');
  }

  const key = getDestinationCacheKey(decodedJwt, destination.name, isolation);
  const expiresIn = first(destination.authTokens || [])?.expiresIn;
  const expirationTime = expiresIn
    ? Date.now() + parseInt(expiresIn) * 1000
    : undefined;
  cache.set(key, { entry: destination, expires: expirationTime });
}

/**
 * Sets the custom destination cache instance.
 * Call this method with an instance of {@link DestinationCacheInterface} to override the default cache instance set by the SDK.
 *
 * NOTE: This function should be called at the beginning before any calls to either {@link getDestination} or {@link @sap-cloud-sdk/http-client!executeHttpRequest}.
 * @param cache - An instance of {@link DestinationCacheInterface}.
 */
export function setDestinationCache(cache: DestinationCacheInterface): void {
  destinationCache = DestinationCache(cache);
}

/**
 * @internal
 */
export let destinationCache: DestinationCacheType = DestinationCache();
/**
 * Determin the default Isolation strategy if not given as option.
 * @param jwt - JWT to determine the default isolation strategy
 * @returns The isolation strategy based on the JWT. If no JWT is given it defaults to Tenant isolation
 * @internal
 */
export function getDefaultIsolationStrategy(
  jwt: JwtPayload | undefined
): IsolationStrategy {
  return jwt && userId(jwt)
    ? IsolationStrategy.Tenant_User
    : IsolationStrategy.Tenant;
}
