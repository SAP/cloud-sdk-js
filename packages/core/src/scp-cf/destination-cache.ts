/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import { MapType } from '@sap-cloud-sdk/util';
import { tenantId, userId } from '../util';
import { Cache, IsolationStrategy } from './cache';
import { DestinationsByType } from './destination-accessor-types';
import { Destination } from './destination-service-types';

const DestinationCache = (cache: Cache<Destination>) => ({
  retrieveDestinationFromCache: (
    decodedJwt: MapType<any>,
    name: string,
    isolation: IsolationStrategy
  ): Destination | undefined =>
    cache.get(getDestinationCacheKey(decodedJwt, name, isolation)),
  cacheRetrievedDestinations: (
    decodedJwt: MapType<any>,
    retrievedDestinations: DestinationsByType,
    isolation: IsolationStrategy
  ): void => {
    [
      ...retrievedDestinations.instance,
      ...retrievedDestinations.subaccount
    ].forEach(destination => {
      if (!destination.name) {
        throw new Error('The destination name is undefined.');
      }

      const key = getDestinationCacheKey(
        decodedJwt,
        destination.name,
        isolation
      );
      cache.set(key, destination);
    });
  },
  clear: () => {
    cache.clear();
  },
  getCacheInstance: () => cache
});

/**
 * Calculates a cache key based on the jwt and destination name for the given isolation strategy.
 * Cache keys for strategies are non-overlapping, i.e. using a cache key for strategy [[IsolationStrategy.Tenant]]
 * will not result in a cache hit for a destination that has been cached with strategy [[IsolationStrategy.Tenant_User]].
 *
 * @param decodedJwt - The decoded JWT of the current request.
 * @param destinationName - The name of the destination.
 * @param isolationStrategy - The strategy used to isolate cache entries.
 * @returns The cache key.
 * @hidden
 */
export function getDestinationCacheKey(
  decodedJwt: MapType<any>,
  destinationName: string,
  isolationStrategy: IsolationStrategy
): string {
  switch (isolationStrategy) {
    case IsolationStrategy.No_Isolation:
      return `::${destinationName}`;
    case IsolationStrategy.Tenant_User:
      return `${tenantId(decodedJwt)}:${userId(decodedJwt)}:${destinationName}`;
    case IsolationStrategy.User:
      return `:${userId(decodedJwt)}:${destinationName}`;
    default:
      return `${tenantId(decodedJwt)}::${destinationName}`;
  }
}

export const destinationCache = DestinationCache(
  new Cache<Destination>({ hours: 0, minutes: 5, seconds: 0 })
);
