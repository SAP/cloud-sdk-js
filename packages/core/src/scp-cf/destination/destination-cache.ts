import { tenantId, userId } from '../../util';
import { Cache, IsolationStrategy } from '../cache';
import { Destination } from './destination-service-types';
import { DestinationsByType } from './destination-accessor-types';

const DestinationCache = (cache: Cache<Destination>) => ({
  retrieveDestinationFromCache: (
    decodedJwt: Record<string, any>,
    name: string,
    isolation: IsolationStrategy
  ): Destination | undefined =>
    cache.get(getDestinationCacheKey(decodedJwt, name, isolation)),
  cacheRetrievedDestination: (
    decodedJwt: Record<string, any>,
    destination: Destination,
    isolation: IsolationStrategy
  ): void => {
    cacheRetrievedDestination(decodedJwt, destination, isolation, cache);
  },
  cacheRetrievedDestinations: (
    decodedJwt: Record<string, any>,
    retrievedDestinations: DestinationsByType,
    isolation: IsolationStrategy
  ): void => {
    retrievedDestinations.subaccount.forEach(dest =>
      cacheRetrievedDestination(decodedJwt, dest, isolation, cache)
    );
    retrievedDestinations.instance.forEach(dest =>
      cacheRetrievedDestination(decodedJwt, dest, isolation, cache)
    );
  },
  clear: () => {
    cache.clear();
  },
  getCacheInstance: () => cache
});

/**
 * Calculates a cache key based on the jwt and destination name for the given isolation strategy.
 * Cache keys for strategies are non-overlapping, i.e. using a cache key for strategy [[IsolationStrategy.Tenant]]
 * will not result in a cache hit for a destination that has been cached with strategy [[IsolationStrategy.Tenant_User]].
 *
 * @param decodedJwt - The decoded JWT of the current request.
 * @param destinationName - The name of the destination.
 * @param isolationStrategy - The strategy used to isolate cache entries.
 * @returns The cache key.
 * @hidden
 */
export function getDestinationCacheKey(
  decodedJwt: Record<string, any>,
  destinationName: string,
  isolationStrategy: IsolationStrategy
): string {
  switch (isolationStrategy) {
    case IsolationStrategy.No_Isolation:
      return `::${destinationName}`;
    case IsolationStrategy.Tenant_User:
      return `${tenantId(decodedJwt)}:${userId(decodedJwt)}:${destinationName}`;
    case IsolationStrategy.User:
      return `:${userId(decodedJwt)}:${destinationName}`;
    default:
      return `${tenantId(decodedJwt)}::${destinationName}`;
  }
}

function cacheRetrievedDestination(
  decodedJwt: Record<string, any>,
  destination: Destination,
  isolation: IsolationStrategy,
  cache: Cache<Destination>
): void {
  if (!destination.name) {
    throw new Error('The destination name is undefined.');
  }

  const key = getDestinationCacheKey(decodedJwt, destination.name, isolation);
  cache.set(key, destination);
}

export const destinationCache = DestinationCache(
  new Cache<Destination>({ hours: 0, minutes: 5, seconds: 0 })
);
