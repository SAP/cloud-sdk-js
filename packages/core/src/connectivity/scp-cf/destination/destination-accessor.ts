import { VerifyJwtOptions } from '../jwt';
import { sanitizeDestination } from './destination';
import {
  Destination,
  DestinationNameAndJwt,
  DestinationRetrievalOptions,
  isDestinationNameAndJwt
} from './destination-service-types';
import { searchEnvVariablesForDestination } from './destination-from-env';
import { searchServiceBindingForDestination } from './destination-from-vcap';
import { getDestinationFromDestinationService } from './destination-from-service';
import { DestinationAccessorOptions } from './destination-accessor-types';

/**
 * Returns the parameter if it is a destination, calls [[getDestination]] otherwise (which will try to fetch the destination
 * from the Cloud Foundry destination service).
 *
 * Fetching a destination requires:
 * - a binding to exactly one XSUAA service instance with service plan "application"
 * - a binding to a destination service instance
 *
 * If either of the prerequisites is not met or one of the services returns an error, this function will either throw an error or return a promise that rejects.
 * @param destination - A destination or the necessary parameters to fetch one.
 * @param options - Caching options by fetching destination.
 * @returns A promise resolving to the requested destination on success.
 */
export async function useOrFetchDestination(
  destination: Destination | DestinationNameAndJwt,
  options: DestinationOptions = {}
): Promise<Destination | null> {
  return isDestinationNameAndJwt(destination)
    ? getDestination(destination.destinationName, {
        userJwt: destination.jwt,
        ...options
      })
    : sanitizeDestination(destination);
}

export type DestinationOptions = DestinationAccessorOptions &
  DestinationRetrievalOptions &
  VerifyJwtOptions;

/**
 * @deprecated Since v1.0.1. Use [[getDestination]] instead.
 *
 * Retrieves a destination with the given name from the Cloud Foundry destination service.
 * Returns `null`, if no destination can be found.
 * Requires the following service bindings: destination, XSUAA
 * By default, selects subscriber over provider and instance over subaccount destinations.
 *
 * If the destinations are read from the environment, the jwt will be ignored.
 * @param name - The name of the destination to be retrieved.
 * @param options - The options of the fetching query of the destination that include the JWT of the current request and the strategy for selecting a destination.
 * @returns A promise returning the requested destination on success.
 */
export async function getDestinationOptions(
  name: string,
  options: DestinationOptions = {}
): Promise<Destination | null> {
  return getDestination(name, options);
}

/**
 * Builds a destination from one of three sources (in the given order):
 * - from the environment variable "destinations"
 * - from service bindings
 * - from the destination service
 *
 * If you want to get a destination only from a specific source, use the corresponding function directly
 *  (`getDestinationFromEnvByName`, `destinationForServiceBinding`, `getDestinationFromDestinationService`).
 * @param name - The name of the destination to be retrieved.
 * @param options - Configuration for how to retrieve destinations from the destination service.
 * @returns A promise returning the requested destination on success.
 */
export async function getDestination(
  name: string,
  options: DestinationOptions = {}
): Promise<Destination | null> {
  return (
    searchEnvVariablesForDestination(name) ||
    searchServiceBindingForDestination(name) ||
    getDestinationFromDestinationService(name, options)
  );
}
