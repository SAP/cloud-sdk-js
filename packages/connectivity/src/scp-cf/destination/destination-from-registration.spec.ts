import { createLogger } from '@sap-cloud-sdk/util';
import { JwtHeader, JwtPayload } from 'jsonwebtoken';
import base64url from 'base64url';
import {
  mockServiceBindings,
  providerServiceToken,
  subscriberServiceToken,
  subscriberUserJwt,
  unmockDestinationsEnv,
  xsuaaBindingMock
} from '../../../../../test-resources/test/test-util';
import {
  DestinationWithName,
  registerDestination,
  registerDestinationCache,
  searchRegisteredDestination
} from './destination-from-registration';
import { getDestination } from './destination-accessor';
import { IsolationStrategy } from './destination-cache';

const testDestination: DestinationWithName = {
  name: 'RegisteredDestination',
  url: 'https://example.com'
};

const destinationWithForwarding: DestinationWithName = {
  forwardAuthToken: true,
  url: 'https://mys4hana.com',
  name: 'FORWARD-TOKEN-DESTINATION'
};

describe('register-destination', () => {
  beforeAll(() => {
    mockServiceBindings();
  });

  afterEach(() => {
    registerDestinationCache.clear();
    unmockDestinationsEnv();
  });

  it('registers destination and retrieves it', async () => {
    await registerDestination(testDestination);
    const actual = await getDestination({
      destinationName: testDestination.name
    });
    expect(actual).toEqual(testDestination);
  });

  it('registers destination and retrieves it with JWT', async () => {
    await registerDestination(testDestination, { jwt: providerServiceToken });
    const actual = await getDestination({
      destinationName: testDestination.name,
      jwt: providerServiceToken
    });
    expect(actual).toEqual(testDestination);
  });

  it('returns undefined if destination key is not found', async () => {
    const actual = await searchRegisteredDestination({
      destinationName: 'Non-existing-destination'
    });
    expect(actual).toBeNull();
  });

  it('caches with tenant-isolation if no JWT is given', async () => {
    await registerDestination(testDestination);
    await expect(
      registerDestinationCache
        .getCacheInstance()
        .hasKey(
          `${xsuaaBindingMock.credentials.subaccountid}::RegisteredDestination`
        )
    ).resolves.toBe(true);
  });

  it('caches with tenant isolation if JWT does not contain user-id', async () => {
    await registerDestination(testDestination, { jwt: subscriberServiceToken });
    await expect(
      registerDestinationCache
        .getCacheInstance()
        .hasKey('subscriber::RegisteredDestination')
    ).resolves.toBe(true);
  });

  it('caches with tenant-user-isolation if JWT is given', async () => {
    await registerDestination(testDestination, { jwt: subscriberUserJwt });
    await expect(
      registerDestinationCache
        .getCacheInstance()
        .hasKey('user-sub:subscriber:RegisteredDestination')
    ).resolves.toBe(true);
  });

  it('cache if tenant if you want', async () => {
    await registerDestination(testDestination, {
      jwt: subscriberUserJwt,
      isolationStrategy: IsolationStrategy.Tenant
    });
    await expect(
      registerDestinationCache
        .getCacheInstance()
        .hasKey('subscriber::RegisteredDestination')
    ).resolves.toBe(true);
  });

  it('caches with unlimited time', async () => {
    jest.useFakeTimers('modern');
    registerDestination(testDestination);
    const minutesToExpire = 9999;
    // Shift time to expire the set item
    jest.advanceTimersByTime(60000 * minutesToExpire);
    const actual = await getDestination({
      destinationName: testDestination.name
    });
    expect(actual).toEqual(testDestination);
  });

  it('adds proxy to registered destination', async () => {
    process.env['https_proxy'] = 'some.http.com';
    registerDestination(testDestination);
    const actual = await getDestination({
      destinationName: testDestination.name
    });
    expect(actual?.proxyConfiguration?.host).toEqual('some.http.com');
    delete process.env['https_proxy'];
  });

  it('adds the auth token if forwardAuthToken is enabled', async () => {
    registerDestination(destinationWithForwarding);
    const jwtPayload: JwtPayload = {
      exp: 1234,
      zid: xsuaaBindingMock.credentials.subaccountid
    };
    const jwtHeader: JwtHeader = { alg: 'HS256' };

    const payloadEncoded = base64url(JSON.stringify(jwtPayload));
    const headerEncoded = base64url(JSON.stringify(jwtHeader));

    const fullToken = `${headerEncoded}.${payloadEncoded}.SomeHash`;
    const actual = await getDestination({
      destinationName: 'FORWARD-TOKEN-DESTINATION',
      jwt: fullToken,
      isolationStrategy: IsolationStrategy.Tenant
    });
    expect(actual?.authTokens![0].expiresIn).toEqual('1234');
    expect(actual?.authTokens![0].value).toEqual(fullToken);
    expect(actual?.authTokens![0].http_header.value).toEqual(
      `Bearer ${fullToken}`
    );
  });

  it('warns if forwardAuthToken is enabled but no token provided.', async () => {
    registerDestination(destinationWithForwarding);

    const logger = createLogger('register-destination');
    const warnSpy = jest.spyOn(logger, 'warn');
    await getDestination({ destinationName: 'FORWARD-TOKEN-DESTINATION' });
    expect(warnSpy).toHaveBeenCalledWith(
      expect.stringMatching(
        /Option 'forwardAuthToken' was set on destination but no token was provided to forward./
      )
    );
  });

  it('infos if registered destination is retrieved.', async () => {
    registerDestination(testDestination);

    const logger = createLogger('register-destination');
    const infoSpy = jest.spyOn(logger, 'info');
    await getDestination({ destinationName: testDestination.name });
    expect(infoSpy).toHaveBeenCalledWith(
      expect.stringMatching(
        /Successfully retrieved destination '\w+' from registered destinations./
      )
    );
  });
});

describe('register-destination without xsuaa binding', () => {
  beforeAll(() => {
    mockServiceBindings(undefined, false);
  });

  afterEach(async () => {
    await registerDestinationCache.clear();
    unmockDestinationsEnv();
  });

  it('registers destination and retrieves it with JWT', async () => {
    await registerDestination(testDestination, { jwt: providerServiceToken });
    const actual = await getDestination({
      destinationName: testDestination.name,
      jwt: providerServiceToken
    });
    expect(actual).toEqual(testDestination);
  });

  it('throws an error when no JWT is provided', async () => {
    await expect(() =>
      registerDestination(testDestination)
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});
