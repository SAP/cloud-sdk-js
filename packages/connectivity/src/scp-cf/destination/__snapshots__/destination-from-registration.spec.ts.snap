// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`register-destination without xsuaa binding throws an error when no JWT is provided 1`] = `"No binding to an XSUAA service instance found. Please make sure to bind an instance of the XSUAA service to your application."`;
